# Unit 12 - Tales from the Crypto

---


## 1. Sentiment Analysis

Use the [newsapi](https://newsapi.org/) to pull the latest news articles for Bitcoin and Ethereum and create a DataFrame of sentiment scores for each coin.

Use descriptive statistics to answer the following questions:
1. Which coin had the highest mean positive score?
2. Which coin had the highest negative score?
3. Which coin had the highest positive score?

In [1]:
# Initial imports
import os
import pandas as pd
from dotenv import load_dotenv
load_dotenv()
import nltk as nltk
nltk.download('vader_lexicon')
from nltk.sentiment.vader import SentimentIntensityAnalyzer
analyzer = SentimentIntensityAnalyzer()

%matplotlib inline

[nltk_data] Downloading package vader_lexicon to
[nltk_data]     /Users/jcmignon/nltk_data...
[nltk_data]   Package vader_lexicon is already up-to-date!


In [2]:
# Read your api key environment variable
api_key = os.getenv("NEWS_API_KEY")

In [3]:
# Create a newsapi client
from newsapi import NewsApiClient
newsapi = NewsApiClient(api_key=api_key)

In [36]:
# Fetch the Bitcoin news articles
btc_news1 = pd.DataFrame(newsapi.get_everything(q="bitcoin", language ="en",sort_by='relevancy',page = 1)['articles'])
btc_news2 = pd.DataFrame(newsapi.get_everything(q="bitcoin", language ="en",sort_by='relevancy',page = 2)['articles'])
btc_news3 = pd.DataFrame(newsapi.get_everything(q="bitcoin", language ="en",sort_by='relevancy',page = 3)['articles'])
btc_news4 = pd.DataFrame(newsapi.get_everything(q="bitcoin", language ="en",sort_by='relevancy',page = 4)['articles'])
btc_news5 = pd.DataFrame(newsapi.get_everything(q="bitcoin", language ="en",sort_by='relevancy',page = 5)['articles'])

btc_news = pd.concat([btc_news1,btc_news2,btc_news3,btc_news4,btc_news5])
btc_news

source  \
0   {'id': 'engadget', 'name': 'Engadget'}   
1      {'id': None, 'name': 'Gizmodo.com'}   
2   {'id': 'mashable', 'name': 'Mashable'}   
3   {'id': 'bbc-news', 'name': 'BBC News'}   
4      {'id': None, 'name': 'Gizmodo.com'}   
..                                     ...   
15    {'id': 'reuters', 'name': 'Reuters'}   
16    {'id': 'reuters', 'name': 'Reuters'}   
17    {'id': 'reuters', 'name': 'Reuters'}   
18    {'id': 'reuters', 'name': 'Reuters'}   
19    {'id': 'reuters', 'name': 'Reuters'}   

                                               author  \
0   https://www.engadget.com/about/editors/saqib-shah   
1                                          Matt Novak   
2                                      Stan Schroeder   
3                    https://www.facebook.com/bbcnews   
4                                       Alyse Stanley   
..                                                ...   
15                                            Reuters   
16                                          Huw Jones   
17                                       Iain Withers   
18                                            Reuters   
19                                         Tom Wilson   

                                                title  \
0   El Salvador becomes the first country to appro...   
1   El Salvador Becomes First Country to Recognize...   
2   Elon Musk says Tesla will resume Bitcoin purch...   
3   Bitcoin: El Salvador makes cryptocurrency lega...   
4   Miami's Bitcoin Conference May Be the Latest C...   
..                                                ...   
15  Sotheby's diamond auction marks another bitcoi...   
16  Bank regulators plan 'conservative' capital ru...   
17  Dollar holds gains after Fed boost; bitcoin tu...   
18  Square to invest $5 mln in Blockstream's solar...   
19  EXCLUSIVE El Salvador bitcoin transfers soar, ...   

                                          description  \
0   El Salvador has voted to adopt Bitcoin as lega...   
1   El Salvador has become the first country in th...   
2   It's all about clean energy, it seems. \nElon ...   
3   It is the first country in the world to make t...   
4   Several crypto fans that descended on Miami, F...   
..                                                ...   
15  A rare pear-shaped diamond that is expected to...   
16  Banks must set aside enough capital to cover l...   
17  The dollar held on to the previous week's gain...   
18  Blockchain technology company Blockstream Mini...   
19  Small transfers of bitcoin to El Salvador jump...   

                                                  url  \
0   https://www.engadget.com/el-salvador-bitcoin-l...   
1   https://gizmodo.com/el-salvador-becomes-first-...   
2   https://mashable.com/article/tesla-bitcoin-pur...   
3   https://www.bbc.co.uk/news/world-latin-america...   
4   https://gizmodo.com/miamis-bitcoin-conference-...   
..                                                ...   
15  https://www.reuters.com/technology/sothebys-di...   
16  https://www.reuters.com/business/bank-regulato...   
17  https://www.reuters.com/business/dollar-holds-...   
18  https://www.reuters.com/technology/square-inve...   
19  https://www.reuters.com/business/finance/exclu...   

                                           urlToImage           publishedAt  \
0   https://s.yimg.com/os/creatr-uploaded-images/2...  2021-06-09T12:04:40Z   
1   https://i.kinja-img.com/gawker-media/image/upl...  2021-06-09T10:00:00Z   
2   https://mondrian.mashable.com/2021%252F06%252F...  2021-06-14T07:15:49Z   
3   https://ichef.bbci.co.uk/news/1024/branded_new...  2021-06-09T08:27:58Z   
4   https://i.kinja-img.com/gawker-media/image/upl...  2021-06-11T00:45:00Z   
..                                                ...                   ...   
15  https://www.reuters.com/resizer/WRdJ_6_WGCwTd1...  2021-06-21T15:16:00Z   
16  https://www.reuters.com/resizer/nEtevxF8w62kQ6...  2021-06-10T10:08:00Z   
17  https://www.reuters.com/resi

In [16]:
# Fetch the Ethereum news articles
eth_news = newsapi.get_everything(q="ethereum", language ="en",sort_by='relevancy')

eth_news1 = pd.DataFrame(newsapi.get_everything(q="ethereum", language ="en",sort_by='relevancy',page = 1)['articles'])
eth_news2 = pd.DataFrame(newsapi.get_everything(q="ethereum", language ="en",sort_by='relevancy',page = 2)['articles'])
eth_news3 = pd.DataFrame(newsapi.get_everything(q="ethereum", language ="en",sort_by='relevancy',page = 3)['articles'])
eth_news4 = pd.DataFrame(newsapi.get_everything(q="ethereum", language ="en",sort_by='relevancy',page = 4)['articles'])
eth_news5 = pd.DataFrame(newsapi.get_everything(q="ethereum", language ="en",sort_by='relevancy',page = 5)['articles'])

eth_news = pd.concat([eth_news1,eth_news2,eth_news3,eth_news4,eth_news5])
eth_news.head()

source                   author  \
0  {'id': 'mashable', 'name': 'Mashable'}             Joseph Green   
1  {'id': 'mashable', 'name': 'Mashable'}               Tim Marcin   
2    {'id': 'reuters', 'name': 'Reuters'}                  Reuters   
3    {'id': None, 'name': 'Entrepreneur'}  Entrepreneur en Español   
4    {'id': None, 'name': 'Slashdot.org'}                   BeauHD   

                                               title  \
0  This blockchain development course bundle is o...   
1               Classic memes that have sold as NFTs   
2  Ethereum extends gains to rise 8%; bitcoin fir...   
3  Ethereum creator Vitalik Buterin made more tha...   
4  Norton 360 Antivirus Now Lets You Mine Ethereu...   

                                         description  \
0  TL;DR: The Cryptocurrency with Ethereum and So...   
1  It wasn't long ago that your average person ha...   
2  Cryptocurrency Ethereum extended gains to rise...   
3  Ethereum billionaire founder Vitalik Buterin b...   
4  NortonLifelock has added the ability to mine E...   

                                                 url  \
0  https://mashable.com/uk/shopping/june-17-crypt...   
1  https://mashable.com/article/classic-memes-sol...   
2  https://www.reuters.com/technology/ethereum-ri...   
3        https://www.entrepreneur.com/article/374264   
4  https://slashdot.org/story/21/06/03/0358213/no...   

                                          urlToImage           publishedAt  \
0  https://mondrian.mashable.com/2021%252F06%252F...  2021-06-17T04:05:00Z   
1  https://mondrian.mashable.com/2021%252F06%252F...  2021-06-20T19:28:07Z   
2  https://www.reuters.com/resizer/yeaYVdBQ4_EGCb...  2021-05-31T12:17:00Z   
3  https://assets.entrepreneur.com/content/3x2/20...  2021-06-10T19:12:00Z   
4        https://a.fsdn.com/sd/topics/bitcoin_64.png  2021-06-03T10:00:00Z   

                                             content  
0  TL;DR: The Cryptocurrency with Ethereum and So...  
1  It wasn't long ago that your average person ha...  
2  A representation of virtual currency Ethereum ...  
3  This article was translated from our Spanish e...  
4  This new mining feature is called 'Norton Cryp...

In [22]:
# Create the Bitcoin sentiment scores DataFrame
btc_sentiment = []

for content in btc_news['content']:
    sentiment = analyzer.polarity_scores(content)
    compound = sentiment["compound"]
    pos = sentiment["pos"]
    neu = sentiment["neu"]
    neg = sentiment["neg"]
        
    btc_sentiment.append({
            "text": content,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu   
        })

# Create DataFrame
btc_df = pd.DataFrame(btc_sentiment)

# Reorder DataFrame columns
cols = ["text", "compound", "positive", "negative", "neutral"]
btc_df = btc_df[cols]

btc_df.head()

text  compound  positive  \
0  El Salvador's President Nayib Bukele has made ...    0.8402     0.282   
1  El Salvador has become the first country in th...    0.1280     0.043   
2  It's all about clean energy, it seems. \r\nElo...    0.6908     0.169   
3  image captionThe move means bitcoin will be ac...    0.2732     0.060   
4  Several crypto fans that descended on Miami, F...    0.5574     0.107   

   negative  neutral  
0       0.0    0.718  
1       0.0    0.957  
2       0.0    0.831  
3       0.0    0.940  
4       0.0    0.893

In [24]:
# Create the Bitcoin sentiment scores DataFrame
eth_sentiment = []

for content in eth_news['content']:
    sentiment = analyzer.polarity_scores(content)
    compound = sentiment["compound"]
    pos = sentiment["pos"]
    neu = sentiment["neu"]
    neg = sentiment["neg"]
        
    eth_sentiment.append({
            "text": content,
            "compound": compound,
            "positive": pos,
            "negative": neg,
            "neutral": neu   
        })

# Create DataFrame
eth_df = pd.DataFrame(eth_sentiment)

# Reorder DataFrame columns
cols = ["text", "compound", "positive", "negative", "neutral"]
eth_df = eth_df[cols]

eth_df.head()

text  compound  positive  \
0  TL;DR: The Cryptocurrency with Ethereum and So...     0.000       0.0   
1  It wasn't long ago that your average person ha...    -0.296       0.0   
2  A representation of virtual currency Ethereum ...     0.000       0.0   
3  This article was translated from our Spanish e...    -0.340       0.0   
4  This new mining feature is called 'Norton Cryp...     0.000       0.0   

   negative  neutral  
0     0.000    1.000  
1     0.061    0.939  
2     0.000    1.000  
3     0.066    0.934  
4     0.000    1.000

In [25]:
# Describe the Bitcoin Sentiment
btc_df.describe()

compound    positive    negative     neutral
count  100.000000  100.000000  100.000000  100.000000
mean     0.059177    0.046790    0.029330    0.923880
std      0.351163    0.063604    0.044144    0.072482
min     -0.778300    0.000000    0.000000    0.718000
25%     -0.081925    0.000000    0.000000    0.871000
50%      0.000000    0.000000    0.000000    0.929000
75%      0.323650    0.083250    0.062250    1.000000
max      0.840200    0.282000    0.178000    1.000000

In [26]:
# Describe the Ethereum Sentiment
eth_df.describe()

compound    positive    negative     neutral
count  100.000000  100.000000  100.000000  100.000000
mean     0.128436    0.054930    0.024120    0.920970
std      0.347841    0.066266    0.042622    0.076793
min     -0.750600    0.000000    0.000000    0.655000
25%      0.000000    0.000000    0.000000    0.881750
50%      0.000000    0.041000    0.000000    0.927500
75%      0.362175    0.086500    0.057750    1.000000
max      0.848100    0.290000    0.262000    1.000000

### Questions:

Q: Which coin had the highest mean positive score?

*__A: Etheruem had the highest mean positive score, 5.49% vs 4.68% for Bitcoin.__*

Q: Which coin had the highest compound score?

*__A: Ethereum had the highest mean compound score (12.84% vs 5.92% respectively), as well as the highest compound score for an individual article (84.81% vs 84.02% respectively).__*

Q. Which coin had the highest positive score?

*__A: Ethereum had the highest positive score (29% vs 28.2% for ETH)__*

## 2. Natural Language Processing
---
###   Tokenizer

In this section, you will use NLTK and Python to tokenize the text for each coin. Be sure to:
1. Lowercase each word.
2. Remove Punctuation.
3. Remove Stopwords.

In [123]:
from nltk.tokenize import word_tokenize, sent_tokenize
from nltk.corpus import stopwords
from nltk.stem import WordNetLemmatizer, PorterStemmer
from string import punctuation
import re

In [124]:
btc_string=" ".join(btc_news['content'].astype(str))

In [152]:
# Create a list of stopwords
sentence = sent_tokenize(btc_string)
words = word_tokenize(btc_string)
sw = set(stopwords.words('english'))
first_result = [word.lower() for word in words if word.lower() not in sw]

# unsure if the list has to be expanded 

In [135]:
# Instantiate the lemmatizer
lemmatizer = WordNetLemmatizer()

In [156]:
# Complete the tokenizer function
def tokenizer(text):
    """Tokenizes text."""
    
    # Remove the punctuation from text
    regex = re.compile("[^a-zA-Z ]")
    re_clean = regex.sub('', text)

    # Create a tokenized list of the words
    words = word_tokenize(re_clean)    
    
    # Define the stop words
    sw = stopwords.words('english')
    
    # Lemmatize words into root words
    lem = [lemmatizer.lemmatize(word) for word in words]
   
    # Convert the words to lowercase
    output = [word.lower() for word in lem if word.lower() not in sw]
    
    return output

In [154]:
# Create a new tokens column for Bitcoin
btc_token = btc_news['content'].apply(tokenizer)
btc_df['token'] = btc_token.values
btc_df.head()

text  compound  positive  \
0  El Salvador's President Nayib Bukele has made ...    0.8402     0.282   
1  El Salvador has become the first country in th...    0.1280     0.043   
2  It's all about clean energy, it seems. \r\nElo...    0.6908     0.169   
3  image captionThe move means bitcoin will be ac...    0.2732     0.060   
4  Several crypto fans that descended on Miami, F...    0.5574     0.107   

   negative  neutral                                              token  
0       0.0    0.718  [el, salvadors, president, nayib, bukele, ha, ...  
1       0.0    0.957  [el, salvador, ha, become, first, country, wor...  
2       0.0    0.831  [clean, energy, seemselon, musk, tesla, caused...  
3       0.0    0.940  [image, captionthe, move, mean, bitcoin, accep...  
4       0.0    0.893  [several, crypto, fan, descended, miami, flori...

In [155]:
# Create a new tokens column for Ethereum
eth_token = eth_news['content'].apply(tokenizer)
eth_df['token'] = eth_token.values
eth_df.head()

text  compound  positive  \
0  TL;DR: The Cryptocurrency with Ethereum and So...     0.000       0.0   
1  It wasn't long ago that your average person ha...    -0.296       0.0   
2  A representation of virtual currency Ethereum ...     0.000       0.0   
3  This article was translated from our Spanish e...    -0.340       0.0   
4  This new mining feature is called 'Norton Cryp...     0.000       0.0   

   negative  neutral                                              token  
0     0.000    1.000  [tldr, cryptocurrency, ethereum, solidity, blo...  
1     0.061    0.939  [wasnt, long, ago, average, person, clue, nft,...  
2     0.000    1.000  [representation, virtual, currency, ethereum, ...  
3     0.066    0.934  [article, wa, translated, spanish, edition, us...  
4     0.000    1.000  [new, mining, feature, called, norton, crypto,...

---

### NGrams and Frequency Analysis

In this section you will look at the ngrams and word frequency for each coin. 

1. Use NLTK to produce the n-grams for N = 2. 
2. List the top 10 words for each coin. 

In [159]:
from collections import Counter
from nltk import ngrams

In [169]:
btc_string = " ".join(btc_df['token'].astype(str))

In [165]:
# Generate the Bitcoin N-grams where N=2
btc_counts = Counter(ngrams(btc_string, n=2))
print(dict(btc_counts))

{('[', "'"): 100, ("'", 'e'): 76, ('e', 'l'): 66, ('l', "'"): 103, ("'", ','): 1868, (',', ' '): 1868, (' ', "'"): 1868, ("'", 's'): 188, ('s', 'a'): 61, ('a', 'l'): 120, ('l', 'v'): 29, ('v', 'a'): 38, ('a', 'd'): 67, ('d', 'o'): 67, ('o', 'r'): 110, ('r', 's'): 110, ('s', "'"): 195, ("'", 'p'): 125, ('p', 'r'): 76, ('r', 'e'): 298, ('e', 's'): 153, ('s', 'i'): 57, ('i', 'd'): 50, ('d', 'e'): 59, ('e', 'n'): 200, ('n', 't'): 106, ('t', "'"): 164, ("'", 'n'): 29, ('n', 'a'): 43, ('a', 'y'): 80, ('y', 'i'): 12, ('i', 'b'): 13, ('b', "'"): 11, ("'", 'b'): 203, ('b', 'u'): 17, ('u', 'k'): 7, ('k', 'e'): 45, ('l', 'e'): 64, ('e', "'"): 249, ("'", 'h'): 44, ('h', 'a'): 132, ('a', "'"): 51, ("'", 'm'): 106, ('m', 'a'): 55, ("'", 'g'): 39, ('g', 'o'): 6, ('o', 'o'): 6, ('o', 'd'): 9, ('d', "'"): 197, ('r', 'o'): 56, ('o', 'm'): 30, ('m', 'i'): 41, ('i', 's'): 38, ('s', 'e'): 113, ("'", 'a'): 94, ('o', 'p'): 21, ('p', 't'): 61, ('b', 'i'): 139, ('i', 't'): 162, ('t', 'c'): 123, ('c', 'o'): 203

In [17]:
# Generate the Ethereum N-grams where N=2
# YOUR CODE HERE!

In [18]:
# Function token_count generates the top 10 words for a given coin
def token_count(tokens, N=3):
    """Returns the top N tokens from the frequency count"""
    return Counter(tokens).most_common(N)

In [19]:
# Use token_count to get the top 10 words for Bitcoin

# print(dict(btc_counts.most_common(10)))


In [20]:
# Use token_count to get the top 10 words for Ethereum

# print(dict(eth_counts.most_common(10)))


---

### Word Clouds

In this section, you will generate word clouds for each coin to summarize the news for each coin

In [21]:
from wordcloud import WordCloud
import matplotlib.pyplot as plt
plt.style.use('seaborn-whitegrid')
import matplotlib as mpl
mpl.rcParams['figure.figsize'] = [20.0, 10.0]

In [22]:
# Generate the Bitcoin word cloud
big_string = ' '.join(btc_df['token'])
btc_text = process_text(big_string)

wc = WordCloud().generate(btc_text)
plt.imshow(wc)

In [23]:
# Generate the Ethereum word cloud
wc = WordCloud().generate(eth_text)
plt.imshow(wc)

---
## 3. Named Entity Recognition

In this section, you will build a named entity recognition model for both Bitcoin and Ethereum, then visualize the tags using SpaCy.

In [24]:
import spacy
from spacy import displacy

In [25]:
# Download the language model for SpaCy
# !python -m spacy download en_core_web_sm

In [26]:
# Load the spaCy model
nlp = spacy.load('en_core_web_sm')

---
### Bitcoin NER

In [27]:
# Concatenate all of the Bitcoin text together
# YOUR CODE HERE!

In [28]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [29]:
# Render the visualization
# YOUR CODE HERE!

In [30]:
# List all Entities
# YOUR CODE HERE!

---

### Ethereum NER

In [31]:
# Concatenate all of the Ethereum text together
# YOUR CODE HERE!

In [32]:
# Run the NER processor on all of the text
# YOUR CODE HERE!

# Add a title to the document
# YOUR CODE HERE!

In [33]:
# Render the visualization
# YOUR CODE HERE!

In [34]:
# List all Entities
# YOUR CODE HERE!

---